sdk/core/azure-core/inc/azure/core/io/body_stream.hpp

```diff
@@ -282,4 +284,45 @@ namespace Azure { namespace Core { namespace IO {
     int64_t Length() const override;
   };
 
+  /**
+   * @brief A concrete implementation of #Azure::Core::IO::BodyStream that wraps another stream and
+   * reports progress
+   */
+  class ProgressBodyStream : public BodyStream {
+  private:
+    BodyStream* m_bodyStream;
+    std::atomic_int64_t m_position;
+    std::function<void(int64_t position, int64_t length)> m_callback;
+
+  private:
+    size_t OnRead(uint8_t* buffer, size_t count, Azure::Core::Context const& context) override;
+
+  public:
+    /**
+     * @brief Constructs `%ProgressBodyStream` from a %BodyStream.
+     *
+     * @param bodyStream the body stream to wrap
+     *
+     * @param callback the callback method used to report progress back to the caller
+     *
+     * @remark The #Azure::Core::IO::ProgressBodyStream does not own the wrapped stream
+     * and is not responsible for closing / cleaning up resources
+     */
+    ProgressBodyStream(
+        BodyStream* bodyStream,
```

Comment:
@CaseyCarter Do we want this to be a BodyStream* or something else (a reference? const?)? I know we discussed this long before but I forget what we arrived at. The idea is that some BodyStream-derived class object is passed in here so this new ProgressBodyStream object can wrap it.

Reply:
If the ProgressBodyStream will own the inner stream and needs to guarantee that the inner stream will remains alive, then it needs to be a unique_ptr<BodyStream>. The inner stream will be moved and maintained by the ProgressBodyStream.

This is how the ReliableStream was implemented.
See: sdk/storage/azure-storage-common/inc/azure/storage/common/internal/reliable_stream.hpp

By the way, we didn't name it `ReliableBodyStream` but only `ReliableStream`.  But we don't have a guideline for naming the Streams. We have `MemoryBodyStream`, `WinHttpStream`, `NullBodyStream`, `RandomAccessFileBodyStream`.

I would expect the ProgressBodyStream to follow the same design from the ReliableStream and owns the inner Stream.  This follows the decorator pattern we want for the BodyStreams.

I think the ProgressStream would have a bug if we keep the raw pointer because customers could call Read() to the innerStream and the progressStream wouldn't know it. Like:

```cpp
MemoryBodyStream innerStream(...);
ProgressStream progressStream(&innerStream);

progressStream.Read(...); // This would update the inner stream position and record the progress
innerStream.Read(...); This would break the progress records from the progressStream
```

Reply:
> If the ProgressBodyStream will own the inner stream and needs to guarantee that the inner stream will remains alive, then it needs to be a unique_ptr. The inner stream will be moved and maintained by the ProgressBodyStream.

I'm not sure I can agree.
I believe there's a convention in C/C++ world that the caller has the responsibility to keep the object alive when he passes a pointer or reference of the object to a function, which most of the time is true for sync functions. Like in the code below

```C++
MemoryBodyStream innerStream(...);
ProgressStream progressStream(&innerStream);

blobClient.Upload(progressStream);
```

Our customer anyway needs to guarantee the stream passed to `blobClient::Upload` alive before the function returns. Adding another layer of `progressStream` doesn't make the contract more complicated.

Reply:
> I think the ProgressStream would have a bug if we keep the raw pointer because customers could call Read() to the innerStream and the progressStream wouldn't know it. Like:

This is nothing but a variant of reading the same stream from more than one thread at the same time, which we've decided to be undefined behavior. The customer needs to avoid this situation.

Reply:
ProgressStream doesn't own the inner stream; it just borrows it; I think this should be a reference (then null is also not possible and then we don't need the assert.

Reply:
I used the same approach as in Http.hpp and keyvault_pipeline.hpp/.cpp where the body stream is passed as a pointer and the caller is responsible for keeping the stream around while it is used by the respective objects. i remember similar patterns in .net where you wrap one stream with another yet you need to keep the original stream around, closing it causes the wrapped object to report the stream closed.

Reply:
done